### PR TITLE
feat(e2e test): add basic e2e test

### PR DESCRIPTION
**Pull request by matwld**
Add a basic e2e test
- also change return value of deploy to 0 to match other commands

Run the test with `npm run e2e-test`

pr is in too parts because I accidentally pushed an altered head instead of making a a new commit

### CODE REVIEW

File: src/libs/tmdk/package.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
   "scripts": {
     "test": "./node_modules/.bin/jest --silent",
     "integ-test": "./node_modules/.bin/jest --silent -c jest.config.integration.js",
+    "e2e-test": "./node_modules/.bin/jest -c jest.config.e2e.js",
```
Comment (johnnyw-aws) [inline src/libs/tmdk/package.json]: instead of creating a new `e2e-test` can we just udpdate `integ-test`?

also since integ-test was made silent should e2e-test be made silent as well?

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+// Ensure nuke always operates
+prompts.inject(["Y", "Y"]);
+
+test("basic e2e test", async () => {
+  console.log("//////      BEGIN BASIC E2E TEST     //////");
+  let argv2;
+
+  // 0. Get account info
+  console.log("Initializing AWS client.");
+  initDefaultAwsClients({ region: `${constants.region}` });
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: should region be typed to avoid this string interpolation?

File: src/libs/tmdk/package.json (hunk 1 above)
Comment (matwld) [reply src/libs/tmdk/package.json]: We discussed with AppKit engineers that integ tests will be mocked through jest and verify api calls rather than validate state. I feel this test is better labeled as an e2e test. I agree having 3 types of tests is confusing, but I think the classifications are:
Unit tests: verify simple edge cases and method returns
Integ tests: verify api calls and expected behavior
E2E test: simulate user interaction and verify interaction with AWS

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+  const twinMakerRoleArn =
+    "arn:aws:iam::" + accountId + ":role/" + twinMakerRoleName;
+  const workspaceS3BucketName = constants.workspaceId + "-" + accountId;
+  const workspaceS3BucketArn = "arn:aws:s3:::" + workspaceS3BucketName;
+  const scene1S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.scene1FileName;
+  const model1S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.model1FileName;
+  const model2S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.model2FileName;
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: can we use string interpolation to simplify this?

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 2 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: I think I just copied this from somewhere else, should have removed the interpolation and just pulled region directly. Will update

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 3 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: Yes, will update

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+
+  // 1. Clean up pre-existing resources, if any
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: to avoid accidental deletion of a workspace that had issues and would like to debug, can we instead create a new  workspace (e.g. timestamp-suffixed) each time (since we should be fully cleaning up at end anyway)

we can also include something to help fully delete a workspace as you have here (or extend nuke to do so) if we end up with cruft workspaces to cleanup from integ test failures

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    );
+    await aws().s3.send(
+      new DeleteObjectCommand({
+        Bucket: workspaceS3BucketName,
+        Key: constants.scene2FileName,
+      })
+    );
+    await aws().s3.deleteBucket({ Bucket: workspaceS3BucketName });
+  } catch (e) {
+    console.log("S3 bucket does not exist, as expected.");
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: do we need to handle the case where the first file doesn't exist but the rest do?

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    twinMakerPolicyString = twinMakerPolicyString.replace(
+      "s3ArnDelete",
+      workspaceS3BucketArn + "/DO_NOT_DELETE_WORKSPACE_*"
+    );
+    await aws().iam.putRolePolicy({
+      RoleName: twinMakerRoleName,
+      PolicyName: twinMakerPolicyName,
+      PolicyDocument: twinMakerPolicyString,
+    });
+    await delay(10000); // allow role to propagate
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: can we avoid blind waits? (or make TODO to revisit)

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 5 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: No, s3 delete object commands do not throw exception if the object does not exist

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 6 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: Sure, will add TODO to verify role creation

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-iam.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        "s3:GetBucket",
+        "s3:GetObject",
+        "s3:ListBucket",
+        "s3:PutObject",
+        "s3:ListObjects",
+        "s3:ListObjectsV2",
+        "s3:GetBucketLocation",
+      ],
+      Resource: [
+        "s3ArnStar", // replaced with workspace bucket
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-iam.ts]: maybe make these more distinctive as something to be replaced? (e.g. `___PLACEHOLDER_FOR_WORKSPACE_BUCKET___`)

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 4 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: Sure, will add

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-iam.ts (hunk 7 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-iam.ts]: Will do

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-constants.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import * as path from "path";
+
+export const localResourcesDir = path.join(__dirname, "basic-e2e-resources");
+export const workspaceId = "tmdk-e2e-test-workspace";
+export const region = "us-east-1";
+export const tmdkDirectory = "/tmp/e2e-tmdk-test";
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-constants.ts]: just curious if this could make the integ test not work on certain platforms, e.g. Windows (can add to backlog if so)

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+  argv2 = {
+    _: ["init"],
+    $0: "tmdk_local",
+    region: constants.region,
+    "workspace-id": constants.workspaceId,
+    out: constants.tmdkDirectory,
+  } as Arguments<init.Options>;
+  expect(await init.handler(argv2)).toBe(0);
+
+  // 5. Validate tmdk definition
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: would it be simpler to compare against a pre-saved tmdk definition?

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+  scene2JsonResult.nodes.forEach(function (model: { name: string }) {
+    scene2ModelNames.push(model.name);
+  });
+  expect(scene2ModelNames).toContain(
+    constants.model1FileName.replace(".glb", "")
+  );
+  expect(scene2ModelNames).toContain(
+    constants.model2FileName.replace(".glb", "")
+  );
+  // Verify Models
```
Comment (johnnyw-aws) [inline src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: should we instead compare against the `basic-e2e-resources` ? (general comment for anything in S3 we should just be uploading)

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-constants.ts (hunk 8 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-constants.ts]: path.join should use the correct delimiter, but this is a good point for the tmp folder I use for tmdk. I'll update that to use path.join as well
https://nodejs.org/api/path.html

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 9 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: Yes, will add

**Review by johnnyw-aws — COMMENTED** [review]
thanks for adding this! overall looks good to me, just want to confirm some of the areas to see if ways to simplify

**Comment by matwld** [comment]
Makes sense, I definitely lost steam on simplifying parts as I went on, will make changes and push a commit to resolve. Last concern is whether this should be labeled e2e test or integ test. I am fine with either, both make sense it seems only to matter to how granular we want to be with testing classification

File: src/libs/tmdk/package.json (hunk 1 above)
Comment (johnnyw-aws) [reply src/libs/tmdk/package.json]: ok, I think we can delete `integ-test` for now if it means something specific in AppKit and call these "functional-tests" instead

File: src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts (hunk 10 above)
Comment (matwld) [reply src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts]: This is specifically checking whether the scene definition contains the correct model name. I could have used a const to store the model name but it was easier to just strip the file extension from the already saved file name

File: src/libs/tmdk/package.json (hunk 1 above)
Comment (matwld) [reply src/libs/tmdk/package.json]: Got it, will update. So for now our current testing naming is:
Unit tests: verify api calls and edge cases
Functional tests: verify e2e user experience and AWS interaction

**Review by johnnyw-aws — APPROVED** [review]
